### PR TITLE
# 162163231 Users can create articles

### PR DESCRIPTION
**Pull request by kafuuma**
#### What does this PR do?
Implements a feature for users to create articles

#### Description of Task to be completed?
```
- Users  can create articles
- Users can view all articles
- Authors can delete articles
```
#### How should this be manually tested?
```
- Navigate to the home page (/) to view all article
- Click on any article to read a full article
- Authors in view mode can delete/edit/update articles
i.e /delete
     /articles/create
    /articles/edit/:slug
    /articles/:slug

```

#### What are the relevant pivotal tracker stories?
[#162163231](https://www.pivotaltracker.com/story/show/162163231)

##### Home page
home

##### Edit Article
edit

##### View full Article
view article

### CODE REVIEW

File: src/actions/actionTypes.js [hunk 1]
```diff
@@ -1,7 +1,13 @@
...
-    SIGNUPSUCCESS: 'SIGNUPSUCCESS'
-}
+  SIGNUP: "SIGNUP",
+  LOGIN: "LOGIN",
+  SIGNUPFAIL: "SIGNUPFAIL",
+  SIGNUPSUCCESS: "SIGNUPSUCCESS"
+};
 export default actionTypes;
+
+export const CREATE_ARTICLE = "CREATE_ARTICLE";
```
Comment (Kasulejoseph) [inline src/actions/actionTypes.js]: Define all the actions as key-value pairs inside the `actionTypes` object.

**Review by Kasulejoseph — CHANGES_REQUESTED** [review]

File: src/actions/articleActions/ArticleActions.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import {
+  CREATE_ARTICLE,
+  GET_ARTICLES,
+  GET_SINGLE_ARTICLE,
+  EDIT_ARTICLE,
+  DELETE_ARTICLE
+} from "../actionTypes";
+
+const token =
+  "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwibmFtZSI6ImhlbnJ5MSIsImV4cCI6MTU0ODEwMTkzMH0.YBpxZfjBtLfI3pC9AFWB8uM3q2PAVcHTKz94lT62vSQ";
```
Comment (Kasulejoseph) [inline src/actions/articleActions/ArticleActions.js]: You are using a hard coded token.

**Review by Kasulejoseph — CHANGES_REQUESTED** [review]

File: src/actions/articleActions/ArticleActions.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    headers: {
+      "content-type": "application/json",
+      authorization: `Bearer ${token}`
+    }
+  })
+    .then(res => res.json())
+    .then(res => {
+      dispatch({ type: GET_SINGLE_ARTICLE, payload: res.article });
+    })
+    .catch(err => console.log(err));
```
Comment (Kasulejoseph) [inline src/actions/articleActions/ArticleActions.js]: remove the console logs

File: src/actions/articleActions/ArticleActions.spec.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
+import fetchMock from "fetch-mock";
```
Comment (Kasulejoseph) [inline src/actions/articleActions/ArticleActions.spec.js]: Action tests should be in the action `__test__` folder

File: src/reducers/articleReducers/ArticleReducer.spec.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
+import ArticleReducer from "./ArticleReducer";
```
Comment (Kasulejoseph) [inline src/reducers/articleReducers/ArticleReducer.spec.js]: reducer tests should also be placed in the reducer `__tests__` folder, that is the convention we agreed to.

**Review by SokoPaulSokool — APPROVED** [review]

**Review by Kasulejoseph — APPROVED** [review]

File: src/actions/articleActions/ArticleActions.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  })
+    .then(res => res.json())
+    .then(res => {
+      dispatch({ type: CREATE_ARTICLE, payload: res.article });
+    })
+    .catch(error => error);
+};
+
+const getArticles = () => dispatch => {
+  return fetch(`${HRK_URL}/articles`, {
```
Comment (bibangamba) [inline src/actions/articleActions/ArticleActions.js]: this line is going to fail on Heroku. please change it to `${HRK_URL}articles` and add the trailing slash in your `.env` file.
Your current statement for the URL will evaluate to `https://ah-backend-athena-staging.herokuapp.com/admin/api//articles` - note the `//` after api
 Also, please use meaningful variable names (even if they are long). `BACKEND_URL` or `API_URL` would be better options to `HRK_URL`

**Review by bibangamba — CHANGES_REQUESTED** [review]

File: src/reducers/articleReducers/ArticleReducer.spec.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+const initialState = {
+  article: {},
+  articles: {},
+  view_article: {},
+  edit_article: {},
+  delete_article: {}
+};
+
+describe("test signup reducer", () => {
```
Comment (Baalmart) [inline src/reducers/articleReducers/ArticleReducer.spec.js]: Could you look into this?, should there be a signup reducer in this articles PR?

File: src/actions/articleActions/ArticleActions.js (hunk 6 above)
Comment (kafuuma) [reply src/actions/articleActions/ArticleActions.js]: Thanks @bibangamba for this feedback. I am running the application an it works well with hweroku. We need to meet an checkout this together I think

**Review by joshtrigger — APPROVED** [review]

**Review by kafuuma — COMMENTED** [review]
I have looked into  and corrected it thanks

File: src/actions/actionTypes.js (hunk 1 above)
Comment (JaymesKat) [reply src/actions/actionTypes.js]: @kafuuma Could you address this?

File: src/views/LoginView/Login.jsx [hunk 8]
```diff
@@ -1,24 +1,24 @@
...
-    super(props)
+  constructor(props) {
+    super(props);
     this.state = {
       email: "",
       password: "",
       errors: ""
-    }
+    };
     this.handleChange = this.handleChange.bind(this);
```
Comment (JaymesKat) [inline src/views/LoginView/Login.jsx]: Use arrow function syntax so that you don't need to bind them in the constructor.

**Review by bibangamba — APPROVED** [review]

File: src/components/articles/articleHeader/ArticleHeader.jsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import PropTypes from "prop-types";
+import { MDBContainer, MDBRow, MDBCol } from "mdbreact";
+import "./ArticleHeader.scss";
+import ArticleOptions from "../articleOptions/ArticleOptions";
+
+const ArticleHeader = props => {
+  const { article, canModify, handleDelete, handleLink } = props;
+  return (
+    <MDBContainer className="mt-5 ">
+      <MDBRow>
```
Comment (JaymesKat) [inline src/components/articles/articleHeader/ArticleHeader.jsx]: You have both the ArticleHeader.jsx and NavBar.jsx which is a duplication of work.

**Review by JaymesKat — COMMENTED** [review]
Rename the folder `articles` under `views` to `Articles`

**Review by JaymesKat — APPROVED** [review]
LGTM

**Pull request merged**